File: packages/service/src/domain/commands/command-executor-service.ts
import { ExecutionContext } from '@service/domain/commands/execution-context';
import { Command } from '@service/domain/commands/model/command';
import { CommandProcessingResponse, CommandProcessingResponseBuilder } from '@service/domain/commands/model/command-processing-response';
import { CommandRegistryService } from '@service/domain/commands/registry/command-registry-service';
import { TextCommandParser } from '@service/domain/commands/text-command-parser';
import { Configuration } from '@service/infrastructure/configuration';
import { Logger } from '@service/infrastructure/logger';
import { Service } from 'typedi';

@Service()
class CommandExecutorService {
  private static logger = new Logger('command-executor-service');

  constructor(
    private commandRegistryService: CommandRegistryService,
    private textCommandParser: TextCommandParser,
    private configuration: Configuration,
  ) { }

  async processCommand(command: Command, context: ExecutionContext): Promise<CommandProcessingResponse> {
    const commandDefinition = this.commandRegistryService.getRegistry().get(command.key);
    if (!commandDefinition) return CommandExecutorService.commandDoesNotExistResponse;

    try {
      return await commandDefinition.execute(command, context);
    } catch (e) {
      const errorMessage = (e as Error).message;
      CommandExecutorService.logger.error(errorMessage);
      return new CommandProcessingResponseBuilder()
        .fromMultilineMarkdown(
          errorMessage,
          '',
          `For more usage information checkout \`${this.configuration.COMMAND_PROMPT} help ${commandDefinition.key}\``,
        )
        .build();
    }
  }

  async processFromText(textCommand: string, context: ExecutionContext): Promise<CommandProcessingResponse> {
    const command = this.textCommandParser.parseTextToCommand(textCommand);
    if (!command) return CommandExecutorService.commandDoesNotExistResponse;

    return this.processCommand(command, context);
  }

  private static get commandDoesNotExistResponse(): CommandProcessingResponse {
    return new CommandProcessingResponseBuilder()
      .fromMarkdown('Komenda nie istnieje')
      .build();
  }
}

export { CommandExecutorService };

File: service/src/main/kotlin/xyz/lebkuchenfm/domain/commands/CommandExecutorService.kt
package xyz.lebkuchenfm.domain.commands

import com.github.michaelbull.result.getOrElse
import com.github.michaelbull.result.map
import io.github.oshai.kotlinlogging.KotlinLogging
import xyz.lebkuchenfm.domain.commands.model.Command
import xyz.lebkuchenfm.domain.commands.model.CommandProcessingResult

private val logger = KotlinLogging.logger {}

class CommandExecutorService(
    private val parser: TextCommandParser,
    private val registry: CommandProcessorRegistry,
    private val commandPrompt: String,
) {
    private suspend fun execute(command: Command, context: ExecutionContext): CommandProcessingResult {
        val processor = registry.getProcessorByKey(command.key)
            ?: return CommandProcessingResult.fromMarkdown("Command ${command.key} does not exist.")

        // Check scope requirements
        if (processor.requiredScope != null && processor.requiredScope !in context.scopes) {
            val message = "You don't have permission to use command '${command.key}'. " +
                "Required scope: '${processor.requiredScope.value}'."
            return CommandProcessingResult.fromMarkdown(message)
        }

        return processor.execute(command, context)
    }

    suspend fun executeFromText(text: String, context: ExecutionContext): CommandProcessingResult {
        return parser.parseFromText(text).map { execute(it, context) }.getOrElse { err ->
            logger.error { "Could not parse command '$text'." }
            CommandProcessingResult.fromMultilineMarkdown(
                when (err) {
                    CommandParsingError.IncorrectPrompt -> "Given text is not a command."
                    CommandParsingError.RequiredTokensMissing -> "The command must perform some action."
                },
                "For more information checkout `$commandPrompt help`.",
            )
        }
    }
}
